crates/sui-framework/packages/sui-framework/tests/derived_object_tests.move

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Mysten Labs, Inc.
```

Comment:
Two thing that we should probably test as well:
- That this does not conflict with dynamic fields or dynamic object fields
- We should probably add some `sui-adapter-transactional-tests` for all the weird  transition states we can hit (especially around party objects and TTO)